docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+    Hash []byte
+}
+```
+
+Only a limited set of modules need any other information from the Tendermint block header and specific versions
+of runtime modules tied to specific Tendermint versions can expose this functionality to modules that need it.
+The basic `BlockInfo` service is left fairly generic to insulate the vast majority of other modules from unneeded
+exposure to changes in the Tendermint protocol.
+
+#### Gas Service
+
+The gas `Service` and `Meter` types will be defined in the `cosmossdk.io/core/gas` package.
+
+The gas service handles tracking of gas consumptions at the block and transaction level and also allows
+for overriding the gas meter passed to child calls:
+```go
+package gas
+
+type Service interface {
+	GetMeter(context.Context) Meter
+	GetBlockMeter(context.Context) Meter
+	WithMeter(ctx context.Context, meter Meter) context.Context
+	WithBlockMeter(ctx context.Context, meter Meter) context.Context
+}
+```
+
+#### Inter-module Client
+
+Runtime module implementations should provide an instance of the `appmodule.RootInterModuleClient` interface which
+will allow modules to send messages to and make queries against other modules as described in [ADR 033](./adr-033-protobuf-inter-module-comm.md).
+
+```go
+type InterModuleClient interface {
+    grpc.ClientConnInterface
+	
+	// Address is the ADR-028 address of this client against which messages will be authenticated.
+	Address() []byte
+}
+
+type RootInterModuleClient interface {
+	InterModuleClient
+	
+    DerivedClient(key []byte) InterModuleClient
+}
+```
+
+The `RootInterModuleClient` allows a module to make inter-module calls using its root [ADR 028](./adr-028-public-key-addresses.md)
+address (as returned by the `Address` method).
+
+This client can be used with both `MsgClient`'s and `QueryClient`'s and the protobuf option
+`cosmos.msg.v1.service` will allow the router to detect whether a given service is a query or msg service and
+route requests appropriately.
+
+Module clients that use [ADR 028](./adr-028-public-key-addresses.md) derived module addresses
+can be created using the `DerivedClient` method.
+
+By default, all `Msg`'s registered using the core API will be available for inter-module routing. Queries will only
+be exposed for inter-module routing if their `rpc` methods have the protobuf option `cosmos.query.v1.internal`
+set to true.
+
+The inter-module router will provide a unified interface for sending `Msg`'s and queries
+which ensures all required pre-processing steps are uniformly executed by all modules which need such functionality.
+
+Runtime modules may choose to allow different types of "admin" access to individual modules which may allow for
+bypassing certain authentication checks. This could allow `x/authz` to allow sending messages on behalf of all
+addresses, for example.
+
+#### `appmodule.Service` Bundle Service
+
+To allow modules to declare a dependency on a single "bundle" service, runtime modules should provide an implementation
```

Comment:
Is the intent with this bundled service to make adoption easier?  In the case of manual wiring I'd need to construct one service (constructor defined in `runtime`) for each module is that right?  Might be even more boilerplate than without this uber service.

Reply:
Yes, the intent is to make adoption easier. Not sure it'd be a lot more boilerplate. Depends on how you construct the service, but shouldn't be harder than constructing store keys